src/uuid.cr

```diff
@@ -0,0 +1,277 @@
+  end
+
+  # Creates UUID from bytes, applying *version* and *variant* to the UUID.
+  def initialize(@bytes : StaticArray(UInt8, 16), variant : Variant, version : Version)
+    self.variant = variant
+    self.version = version
+  end
+
+  def initialize(variant : Variant)
+    @bytes = uninitialized UInt8[16]
+    @bytes.to_unsafe.copy_from SecureRandom.random_bytes(16).pointer(16), 16
+
+    variant = variant
+    version = Version::V4
+  end
+
+  # Generates RFC 4122 UUID `variant` with specified `version`.
+  def initialize(@bytes : StaticArray(UInt8, 16), version = Version::V4)
+    case version
+    when Version::V4
+      variant = Variant::RFC4122
+      version = Version::V4
+    else
+      raise ArgumentError.new "Creating #{version} not supported."
+    end
+  end
+
+  # Returns UUID variant.
+  def variant
+    case
+    when @bytes[8] & 0x80 == 0x00
+      Variant::NCS
+    when @bytes[8] & 0xc0 == 0x80
+      Variant::RFC4122
+    when @bytes[8] & 0xe0 == 0xc0
+      Variant::Microsoft
+    when @bytes[8] & 0xe0 == 0xe0
+      Variant::Future
+    else
+      Variant::Unknown
+    end
+  end
+
+  # Sets UUID variant to specified *value*.
+  def variant=(value : Variant)
+    case value
+    when Variant::NCS
+      @bytes[8] = (@bytes[8] & 0x7f)
+    when Variant::RFC4122
+      @bytes[8] = (@bytes[8] & 0x3f) | 0x80
+    when Variant::Microsoft
+      @bytes[8] = (@bytes[8] & 0x1f) | 0xc0
+    when Variant::Future
+      @bytes[8] = (@bytes[8] & 0x1f) | 0xe0
+    else
+      raise ArgumentError.new "Can't set unknown variant."
+    end
+  end
+
+  # Returns version based on RFC4122 format. See also `#variant`.
+  def version
+    case @bytes[6] >> 4
+    when 1 then Version::V1
+    when 2 then Version::V2
+    when 3 then Version::V3
+    when 4 then Version::V4
+    when 5 then Version::V5
+    else        Version::Unknown
+    end
+  end
+
+  # Sets `version`. Doesn't set variant (see `#variant=`).
+  def version=(value : Version)
+    raise ArgumentError.new "Can't set unknown version." if value.unknown?
+    @bytes[6] = (@bytes[6] & 0xf) | (value.to_u8 << 4)
+  end
+
+  def self.empty
+    # new StaticArray(UInt8, 16).new(0_u8)
+    new(StaticArray(UInt8, 16).new(0_u8), UUID::Variant::NCS, UUID::Version::V4)
+  end
+
+  # Returns 16-byte slice.
+  def to_slice
+    Slice(UInt8).new to_unsafe, 16
+  end
+
+  # Returns unsafe pointer to 16-bytes.
+  def to_unsafe
+    @bytes.to_unsafe
+  end
+
+  # Writes hyphenated format string to the *io*.
+  def to_s(io : IO)
+    io << to_s
+  end
+
+  # Returns `true` if `other` string represents the same UUID, `false` otherwise.
+  def ==(other : String)
+    self == UUID.new other
+  end
+
+  # Returns `true` if `other` 16-byte slice represents the same UUID, `false` otherwise.
+  def ==(other : Slice(UInt8))
+    to_slice == other
+  end
+
+  # Returns `true` if `other` static 16 bytes represent the same UUID, `false` otherwise.
+  def ==(other : StaticArray(UInt8, 16))
```

Comment:
I'm similarly unsure about having such weak comparisons with strings, slices, and staticarrays too. I think people really should convert manually to reduce magic.